### PR TITLE
Use a state type for the state field in StateChangedEvent

### PR DESCRIPTION
**Pull request by TTWNO**
Be more strict about the `state` field.

### CODE REVIEW

**Comment by TTWNO** [comment]
I had implemented Default for State, setting Invalid as the default option of none others are available. Unfortunately, this isn't really correct since "invalid" getting sent over the bus, and a state of invalid characters (for example "fire"), should not have the same behaviour.

**Comment by TTWNO** [comment]
Blocked on #91

File: atspi-common/README.md [hunk 1]
```diff
@@ -5,3 +5,19 @@ This crate is meant to only contain the absolute lowest common denominator for l
 therefore, it attempts to be able to compile on basically any architecture (including WASM).
 
 Please use the internal documentation to learn how to use these data structures.
+
+## Compiling for WASM
+
+If you'd like to compile this crate for any architecture that implements `std`, then you have to compile it without any feature flags.
+By default, this crate provides a variety of converison methods for foreign error types into its own event type.
```
Comment (luukvanderduim) [inline atspi-common/README.md:12]: conversions

File: atspi-common/src/state.rs [hunk 2]
```diff
@@ -219,6 +221,122 @@ pub enum State {
 	ReadOnly,
 }
 
+impl From<State> for String {
```
Comment (luukvanderduim) [inline atspi-common/src/state.rs:224]: note you added this conversion to be able to print it elsewhere, then maybe save a conversion and `impl Display`, but you perhaps have other uses in mind as well..

File: atspi-common/src/state.rs [hunk 3]
```diff
@@ -11,10 +12,11 @@ use zvariant::{Signature, Type};
 /// of an accessibility object.
 #[bitflags]
 #[repr(u64)]
-#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq, Eq, Hash, Default)]
```
Comment (luukvanderduim) [inline atspi-common/src/state.rs:15]: Love `Default`. Hash is here to help users storing states in a `HashMap`?

**Review by luukvanderduim — COMMENTED** [review]
Much good stuff.
Though I am unsure about making the error more opaque. I cannot see why that was necessary while on the other hand I could see why a user wants to sift types of error by origin?

File: atspi-common/src/error.rs [hunk 4]
```diff
@@ -30,69 +26,72 @@ pub enum AtspiError {
...
-	UnknownSignal,
+	UnknownSignal(String),
 
 	/// Other errors.
 	Owned(String),
 
 	/// A `zbus` or `zbus::Fdo` error. variant.
 	Zbus(String),
 
-	/// A `zbus_names` error variant
```
[the remark this answers is not in the document]
Comment (TTWNO) [reply atspi-common/src/error.rs:42]: `zvariant` and `zbus_names` error types can not be serialized. My primary issue is that I need to be able to send errors from/to a WASM guest. Eventually, generic reference types should be able to be sent over a WASM FFI boundary, but that is still some months/years out.

I could make a PR to those two repos for a `serde` flag, that would enable serialization of these types. What do you think about that?

File: atspi-common/src/events/mod.rs [hunk 5]
```diff
@@ -555,7 +549,7 @@ impl TryFrom<Event> for AvailableEvent {
 		if let Event::Available(specific_event) = generic_event {
 			Ok(specific_event)
 		} else {
-			Err(AtspiError::Conversion("Invalid type"))
+			Err(AtspiError::InvalidType)
```
[the remark this answers is not in the document]
Comment (TTWNO) [reply atspi-common/src/events/mod.rs:552]: `TypeMismatch` coming up!

File: atspi-common/src/state.rs (hunk 2 above)
Comment (TTWNO) [reply atspi-common/src/state.rs:224]: Technically it is still for conversion. Basically, I just want to infallibly convert to a String for a conversion from a `Message` to a `T: GenericEvent`. I could unwrap a `serde` call here, but just not sure how I feel about that.

File: atspi-common/src/state.rs (hunk 3 above)
Comment (TTWNO) [reply atspi-common/src/state.rs:15]: Using a State as a *key* in a HashMap.

File: atspi-common/src/error.rs [hunk 6]
```diff
@@ -1,11 +1,7 @@
 #[allow(clippy::module_name_repetitions)]
-#[derive(Debug)]
-#[non_exhaustive]
+#[derive(Debug, serde::Serialize, serde::Deserialize, Eq, PartialEq, Clone)]
 /// An error type that can describe atspi and `std` and different `zbus` errors.
 pub enum AtspiError {
-	/// Converting one type into another failure
-	Conversion(&'static str),
-
 	/// When testing on either variant, we might find the we are not interested in.
```
[the remark this answers is not in the document]
Comment (TTWNO) [reply atspi-common/src/error.rs:5]: Not sure what you're commenting on. This was a removed error variant that could not be serialized.

File: atspi-common/src/error.rs [hunk 7]
```diff
@@ -1,11 +1,7 @@
 #[allow(clippy::module_name_repetitions)]
-#[derive(Debug)]
-#[non_exhaustive]
```
[the remark this answers is not in the document]
Comment (TTWNO) [reply atspi-common/src/error.rs:3]: Not sure how `#[non_exhaustive]` got removed. That's on me. Put it back in the latest commit.

File: atspi-common/src/state.rs (hunk 2 above)
Comment (luukvanderduim) [reply atspi-common/src/state.rs:224]: Maybe don't do that. There are still `Message`s which we do not parse to an event. The two DBus messages with "s" signature. (forgot their names)
And if interfaces change, or there are signals we somehow don't yet know of, we do not want to panic unless we absolutely must. We want ATs to handle an error and carry on.

**Pull request closed without merging**